File: apps/api/app/auth/permissions.py
"""Static RBAC permission matrix.

Maps (auth_role, module) -> allowed actions. The matrix is the single source of
truth for what each role can do per IA module; it is consulted by the
`require_permission` FastAPI dependency in `app.auth.rbac`.
"""
from typing import Literal

Role = Literal["admin", "manager", "editor", "drafter", "purchase_officer", "viewer"]
Module = Literal[
    "dashboard",
    "tracking",
    "list",
    "shop_dwgs",
    "isample",
    "orderbook",
    "it_management",
]
Action = Literal["read", "write", "approve", "comment"]

_ALL_MODULES: tuple[str, ...] = (
    "dashboard",
    "tracking",
    "list",
    "shop_dwgs",
    "isample",
    "orderbook",
    "it_management",
)

MATRIX: dict[str, dict[str, set[str]]] = {
    "admin": {m: {"read", "write", "approve", "comment"} for m in _ALL_MODULES},
    "manager": {
        m: {"read", "write", "approve", "comment"}
        for m in _ALL_MODULES
        if m != "it_management"
    }
    | {"it_management": {"read"}},
    "editor": {
        m: {"read", "write", "comment"}
        for m in ("dashboard", "tracking", "list", "shop_dwgs", "isample")
    }
    | {"orderbook": {"read", "comment"}, "it_management": set()},
    "drafter": {
        "dashboard":     {"read"},
        "tracking":      {"read", "write", "approve", "comment"},
        "list":          {"read", "write", "approve", "comment"},
        "shop_dwgs":     {"read", "write", "approve", "comment"},
        "isample":       {"read"},
        "orderbook":     {"read", "write", "approve", "comment"},
        "it_management": set(),
    },
    "purchase_officer": {
        "dashboard": {"read"},
        "tracking": {"read", "comment"},
        "list": {"read"},
        "shop_dwgs": {"read"},
        "isample": {"read"},
        "orderbook": {"read", "write", "approve", "comment"},
        "it_management": set(),
    },
    "viewer": {m: {"read"} for m in _ALL_MODULES if m != "it_management"}
    | {"it_management": set()},
}


def has_permission(role: str, module: str, action: str) -> bool:
    """Return True iff `role` is allowed to perform `action` on `module`."""
    return action in MATRIX.get(role, {}).get(module, set())

File: apps/web/lib/permissions.ts
/**
 * The user shape and the RBAC gate, client-safe.
 *
 * These live here rather than in `lib/session.ts` because that module imports
 * `next/headers`, which makes it server-only — a client component importing
 * `can` from it drags `next/headers` into the browser bundle and fails the
 * build. `lib/session.ts` re-exports everything below, so server code can keep
 * importing from either.
 */

/** IA modules in the RBAC matrix (`_ALL_MODULES` in permissions.py). */
export type Module =
  | "dashboard"
  | "tracking"
  | "list"
  | "shop_dwgs"
  | "isample"
  | "orderbook"
  | "catalog"
  | "cut_floor"
  | "shop_floor"
  | "estimating"
  | "it_management";

export type Action = "read" | "write" | "approve" | "comment";

export type Me = {
  id: number;
  workspace_id: number;
  email: string;
  full_name: string;
  auth_role:
    | "admin"
    | "manager"
    | "editor"
    | "drafter"
    | "estimator"
    | "purchase_officer"
    | "viewer";
  /** `{module: [actions]}` for this user's role, served by `/auth/me`. */
  permissions: Partial<Record<Module, Action[]>>;
};

/**
 * Gate a surface on the RBAC matrix rather than a hardcoded role list, so the
 * web tier can never drift from `apps/api/app/auth/permissions.py`.
 *
 * Fails closed: a null user, or a `me` from a cached response predating the
 * `permissions` field, denies.
 */
export function can(
  me: Me | null | undefined,
  module: Module,
  action: Action,
): boolean {
  return me?.permissions?.[module]?.includes(action) ?? false;
}
